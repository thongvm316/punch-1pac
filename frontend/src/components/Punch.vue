<template>
  <div>
    <div class="punch">
      <span v-if="attendance.attended_at">{{ $t('header.in') }}: {{ attendance.attended_at }}</span>
      <span
        v-if="attendance.attended_at"
        class="mx-2"
      >-</span>
      <span
        v-if="attendance.left_at"
        class="mr-5"
      >{{ $t('header.out') }}: {{ attendance.left_at }}</span>
      <span
        v-else
        class="mr-5"
      >{{ currentTime }}</span>
      <button
        v-if="!attendance.attended_at"
        ref="btnPunchIn"
        class="btn btn-primary mr-5"
        @click="debouncePunchIn()"
      >
        {{ $t('button.common.punchIn') }}
      </button>
      <button
        v-if="attendance.attended_at && !attendance.left_at"
        ref="btnPunchOut"
        class="btn btn-primary mr-5"
        @click="openConfirmDialog()"
      >
        {{ $t('button.common.punchOut') }}
      </button>
    </div>
    <confirm-dialog
      v-if="!attendance.left_at"
      :title="$t('header.punchOutTitle')"
      :delete-object="debouncePunchOut"
      :modal-open.sync="isOpenConfirmDialog"
    >
      <p>{{ $t('header.punchOutConfirm', { at: $moment().format('HH:mm') }) }}</p>
    </confirm-dialog>
  </div>
</template>

<script>
import { mapState, mapActions, mapMutations } from 'vuex'
import confirmDialog from '@/mixins/confirm-dialog'
import { PUNCH_INIT_ATTENDANCE, SET_FLASH_MESSAGE } from '@/store/mutation-types'

export default {
  name: 'Punch',

  mixins: [confirmDialog],

  data() {
    return {
      isPunching: false,
      currentTime: null,
      timer: null
    }
  },

  computed: {
    ...mapState('punch', ['isInited', 'attendance'])
  },

  created() {
    if (!this.isInited) this[PUNCH_INIT_ATTENDANCE](window.initialStates().attendance)
    this.updateCurrentTime()
    this.timer = setInterval(this.updateCurrentTime, 1 * 1000)
  },

  destroyed() {
    clearInterval(this.timer)
  },

  methods: {
    updateCurrentTime() {
      this.currentTime = this.$moment().format('HH:mm:ss')
    },

    ...mapMutations('punch', [PUNCH_INIT_ATTENDANCE]),

    ...mapMutations('flash', [SET_FLASH_MESSAGE]),

    ...mapActions('punch', ['punchIn', 'punchOut']),

    debouncePunchIn() {
      if (this.isPunching) return
      this.isPunching = true
      this.punchIn(this.currentUser.id).then(response => {
        this.isPunching = false
        if (response.data) this[SET_FLASH_MESSAGE]({ message: this.$t('header.punchInSuccess', { at: response.data.attended_at }) })
      })
    },

    debouncePunchOut(attendance = {}) {
      if (this.isPunching) return
      this.isPunching = true
      this.punchOut(this.currentUser.id).then(response => {
        this.isPunching = false
        this.isOpenConfirmDialog = false
        this[SET_FLASH_MESSAGE]({ message: this.$t('header.punchOutSuccess', { at: response.data.left_at }) })
      })
    }
  }
}
</script>
