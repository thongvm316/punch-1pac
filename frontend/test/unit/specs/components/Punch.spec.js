import { shallowMount } from '@vue/test-utils'
import wrapperOps from '../../supports/wrapper'
import Punch from '@/components/Punch'

const fakePunchResponse = {
  in: {
    data: {
      attended_at: '14:30',
      left_at: null
    }
  },
  out: {
    data: {
      attended_at: '14:30',
      left_at: '17:30'
    }
  }
}

const debouncePunchIn = jest.spyOn(Punch.methods, 'debouncePunchIn')
const updateCurrentTime = jest.spyOn(Punch.methods, 'updateCurrentTime')
const punchIn = jest.spyOn(Punch.methods, 'punchIn').mockResolvedValue(fakePunchResponse.in)
const punchOut = jest.spyOn(Punch.methods, 'punchOut').mockResolvedValue(fakePunchResponse.out)
const openConfirmDialog = jest.spyOn(Punch.mixins[0].methods, 'openConfirmDialog')
const PUNCH_INIT_ATTENDANCE = jest.spyOn(Punch.methods, 'PUNCH_INIT_ATTENDANCE')
const SET_FLASH_MESSAGE = jest.spyOn(Punch.methods, 'SET_FLASH_MESSAGE')

describe('PopupChangePassword.vue', () => {
  let wrapper

  afterEach(() => { wrapper.vm.$destroy() })

  describe('when user have not punch in yet', () => {
    const localWrapperOps = {
      ...wrapperOps,
      stubs: {
        ConfirmDialog: true
      },
      computed: {
        isInited () {
          return false
        },
        attendance () {
          return {}
        },
        currentUser() {
          return { id: 0 }
        }
      }
    }
    beforeEach(() => {
      wrapper = shallowMount(Punch, localWrapperOps)
    })

    it('component should render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.isVueInstance()).toBeTruthy()
      expect(updateCurrentTime).toHaveBeenCalled()
      expect(PUNCH_INIT_ATTENDANCE).toHaveBeenCalled()
      expect(wrapper).toMatchSnapshot()
    })

    it('should call debouncePunchIn when user click punch In', async () => {
      wrapper.find({ ref: 'btnPunchIn' }).trigger('click')
      await wrapper.vm.$nextTick()

      expect(debouncePunchIn).toHaveBeenCalled()
      expect(punchIn).toHaveBeenCalledWith(0)
      expect(SET_FLASH_MESSAGE).toHaveBeenCalledWith({ message: 'You punched in at 14:30' })
    })
  })

  describe('when user have punched in but not punched out yet', () => {
    const localWrapperOps = {
      ...wrapperOps,
      stubs: {
        ConfirmDialog: true
      },
      computed: {
        isInited () {
          return false
        },
        attendance () {
          return fakePunchResponse.in.data
        },
        currentUser() {
          return { id: 0 }
        }
      }
    }
    beforeEach(() => {
      wrapper = shallowMount(Punch, localWrapperOps)
    })

    it('component should render correctly', () => {
      expect(wrapper.exists()).toBeTruthy()
      expect(wrapper.isVueInstance()).toBeTruthy()
      expect(updateCurrentTime).toHaveBeenCalled()
      expect(wrapper).toMatchSnapshot()
    })

    it('should openConfirmDialog haveBeenCalled', async () => {
      wrapper.find({ ref: 'btnPunchOut' }).trigger('click')
      await wrapper.vm.$nextTick()

      expect(openConfirmDialog).toHaveBeenCalled()
      expect(wrapper).toMatchSnapshot()
    })

    it('should trigger punchOut', async () => {
      wrapper.vm.debouncePunchOut()
      await wrapper.vm.$nextTick()

      expect(punchOut).toHaveBeenCalledWith(0)
      expect(SET_FLASH_MESSAGE).toHaveBeenCalledWith({ message: 'You punched out at 17:30' })
    })
  })

  describe('when user have punched in and out', () => {
    const localWrapperOps = {
      ...wrapperOps,
      computed: {
        isInited() { return false },
        attendance() { return fakePunchResponse.out.data }
      }
    }

    wrapper = shallowMount(Punch, localWrapperOps)

    it('should not show btn PunchIn, btn PunchOut', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})
