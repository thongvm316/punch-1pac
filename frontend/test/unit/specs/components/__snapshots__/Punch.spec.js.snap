// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupChangePassword.vue when user have not punch in yet component should render correctly 1`] = `
<div>
  <div class="punch">
    <!---->
    <!----> <span class="mr-5">10:39:39</span> <button class="btn btn-primary mr-5">
      Punch In
    </button>
    <!---->
  </div>
  <confirmdialog-stub title="Punch Out" delete-object="function () { [native code] }">
    <p>Do you intent to punch out at 10:39 ?</p>
  </confirmdialog-stub>
</div>
`;

exports[`PopupChangePassword.vue when user have punched in and out should not show btn PunchIn, btn PunchOut 1`] = `
<div>
  <div class="punch"><span>In: 14:30</span> <span class="mx-2">-</span> <span class="mr-5">10:39:39</span>
    <!----> <button class="btn btn-primary mr-5">
      Punch Out
    </button></div>
  <confirmdialog-stub title="Punch Out" delete-object="function () { [native code] }">
    <p>Do you intent to punch out at 10:39 ?</p>
  </confirmdialog-stub>
</div>
`;

exports[`PopupChangePassword.vue when user have punched in but not punched out yet component should render correctly 1`] = `
<div>
  <div class="punch"><span>In: 14:30</span> <span class="mx-2">-</span> <span class="mr-5">10:39:39</span>
    <!----> <button class="btn btn-primary mr-5">
      Punch Out
    </button></div>
  <confirmdialog-stub title="Punch Out" delete-object="function () { [native code] }">
    <p>Do you intent to punch out at 10:39 ?</p>
  </confirmdialog-stub>
</div>
`;

exports[`PopupChangePassword.vue when user have punched in but not punched out yet should openConfirmDialog haveBeenCalled 1`] = `
<div>
  <div class="punch"><span>In: 14:30</span> <span class="mx-2">-</span> <span class="mr-5">10:39:39</span>
    <!----> <button class="btn btn-primary mr-5">
      Punch Out
    </button></div>
  <confirmdialog-stub title="Punch Out" delete-object="function () { [native code] }" modal-open="true">
    <p>Do you intent to punch out at 10:39 ?</p>
  </confirmdialog-stub>
</div>
`;
